### PR TITLE
PLT-445 Add `mustReferencePubKeyOutput` in constraints library

### PR DESCRIPTION
**Pull request by koslambrou**
* Added the `mustReferencePubKeyOutput` constraint in `plutus-ledger-constraints` and `plutus-tx-constraints`
    
* Added a test case for the `mustReferencePubKeyOutput`, but they are not run in CI until we remove our custom transaction validation rules in favor of using fully `cardano-ledger` validation rules
    
* Moved `TxIn` and `TxInType`  from `plutus-ledger-api` into `plutus-ledger` (these types will be removed from `plutus`)

* Added plutus version to `ConsumeScriptAddress` constructor of `TxInType`

* Renamed `Ledger.Index.mkTxInfo` into `mkPV1TxInfo`

* Added reference inputs field in `Ledger.Tx.Internal`

* Added constraints `MustUseOutputCollateral,` `MustSpendScriptOutput` and `MustPayToOtherScript` constraints in order to implement the test case with `plutus-tx-constraints` 

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-contract/test/Spec/Contract/TxConstraints.hs [hunk 1]
```diff
@@ -0,0 +1,259 @@
+{-# LANGUAGE DataKinds             #-}
```
Comment (koslambrou) [inline plutus-contract/test/Spec/Contract/TxConstraints.hs:1]: @james-iohk Feel free to verify that the testing is valid and propose new testing for this constraint.

File: plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V2.hs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        traceIfFalse "Lb" -- "MustPayToOtherScript"
+        $ any checkOutput outs
+    MustHashDatum dvh dv ->
+        traceIfFalse "Lc" -- "MustHashDatum"
+        $ PV2.findDatum dvh scriptContextTxInfo == Just dv
+    MustSatisfyAnyOf xs ->
+        traceIfFalse "Ld" -- "MustSatisfyAnyOf"
+        $ any (all (checkTxConstraint ctx)) xs
+    MustUseOutputAsCollateral _ ->
+        True -- TODO
```
Comment (ak3n) [inline plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V2.hs:155]: Empty todo.

File: plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs [hunk 3]
```diff
@@ -133,6 +133,11 @@ checkTxConstraint ctx@ScriptContext{scriptContextTxInfo} = \case
     MustSatisfyAnyOf xs ->
         traceIfFalse "Ld" -- "MustSatisfyAnyOf"
         $ any (all (checkTxConstraint ctx)) xs
+    MustUseOutputAsCollateral _ ->
+        True -- TODO
```
Comment (ak3n) [inline plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs:137]: Empty todo.

File: plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs [hunk 4]
```diff
@@ -145,7 +150,7 @@ checkTxConstraintFun ScriptContext{scriptContextTxInfo} = \case
                 vh == vh' && valuePred val && datumPred d
             isMatch _ = False
         in
-        traceIfFalse "Le" -- "MustSpendScriptOutputWithMatchingDatumAndValue"
+        traceIfFalse "Lf" -- "MustSpendScriptOutputWithMatchingDatumAndValue"
```
Comment (ak3n) [inline plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs:153]: Hmm, why change instead of introducing a new code for new constraint? This way we break the backward compatibility and it may confuse.

File: plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs [hunk 5]
```diff
@@ -43,7 +43,7 @@ import Plutus.V1.Ledger.Value (leq)
 -- | Does the 'ScriptContext' satisfy the constraints?
 checkScriptContext :: forall i o. ToData o => TxConstraints i o -> ScriptContext -> Bool
 checkScriptContext TxConstraints{txConstraints, txConstraintFuns = TxConstraintFuns txCnsFuns, txOwnInputs, txOwnOutputs} ptx =
-    traceIfFalse "Ld" -- "checkV1ScriptContext failed"
+    traceIfFalse "L!" -- "checkScriptContext failed"
```
Comment (ak3n) [inline plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs:46]: Same as below, not sure it's a good thing to change codes.

File: plutus-tx-constraints/src/Ledger/Tx/Constraints/OffChain.hs [hunk 6]
```diff
@@ -218,4 +265,21 @@ processConstraint = \case
...
+        networkId <- use (P.paramsL . networkIdL)
+        out <- throwLeft ToCardanoError $ C.TxOut
+            <$> C.toCardanoAddressInEra networkId (scriptValidatorHashAddress vlh svhM)
+            <*> C.toCardanoTxOutValue vl
+            <*> pure (C.TxOutDatumInTx C.ScriptDataInBabbageEra (C.toCardanoScriptData (getDatum dv)))
+            <*> pure C.ReferenceScriptNone
+        unbalancedTx . tx . txOuts <>= [ out ]
+
+    P.MustIncludeDatum _ -> do
+        pure () -- TODO
```
Comment (ak3n) [inline plutus-tx-constraints/src/Ledger/Tx/Constraints/OffChain.hs:278]: Empty TODO.

File: plutus-tx-constraints/src/Ledger/Tx/Constraints/OffChain.hs [hunk 7]
```diff
@@ -197,18 +216,46 @@ processConstraint
...
           _ -> throwError (LedgerMkTxError $ P.TxOutRefWrongType txo)
 
-    P.MustPayToPubKeyAddress pk mskh md vl -> do
+    P.MustSpendScriptOutput txo redeemer -> do
+        txout <- lookupTxOutRef txo
+        mscriptTXO <- mapReaderT (mapStateT (mapExcept (first LedgerMkTxError))) $ P.resolveScriptTxOut txout
+        case mscriptTXO of
+            Just ((_, validator), (_, datum), _) -> do
+                txIn <- throwLeft ToCardanoError $ C.toCardanoTxIn txo
+                -- Hardcoded to PlutusV2. Needs to also work with PlutusV1
```
Comment (ak3n) [inline plutus-tx-constraints/src/Ledger/Tx/Constraints/OffChain.hs:233]: Should be `TODO`?

File: plutus-tx-constraints/src/Ledger/Tx/Constraints/OffChain.hs [hunk 8]
```diff
@@ -197,18 +216,46 @@ processConstraint
     -> ReaderT (P.ScriptLookups a) (StateT P.ConstraintProcessingState (Except MkTxError)) ()
 processConstraint = \case
     P.MustSpendPubKeyOutput txo -> do
-        txout <- mapReaderT (mapStateT (mapExcept (first LedgerMkTxError))) $ P.lookupTxOutRef txo
+        txout <- lookupTxOutRef txo
         case txout of
           Tx.PublicKeyChainIndexTxOut {} -> do
               txIn <- throwLeft ToCardanoError $ C.toCardanoTxIn txo
               unbalancedTx . tx . txIns <>= [(txIn, C.BuildTxWith (C.KeyWitness C.KeyWitnessForSpending))]
             --   valueSpentInputs <>= provided _ciTxOutValue
```
Comment (ak3n) [inline plutus-tx-constraints/src/Ledger/Tx/Constraints/OffChain.hs:224]: Can be removed?

File: plutus-ledger/src/Ledger/Typed/Scripts.hs [hunk 9]
```diff
@@ -1,9 +1,116 @@
...
+    Just (ConsumeScriptAddress lang _val re da) -> do
+      rsVal <- checkRedeemer typedValidator re
+      _ <- checkDatum typedValidator da
+      let txOutRef = txInRef txIn
+      case lookupRef txOutRef of
+        Just (txOut, datum) -> do
+          typedOut <- typeScriptTxOutRef @inn typedValidator txOutRef txOut datum
+          pure $ makeTypedScriptTxIn lang typedValidator rsVal typedOut
+        Nothing -> throwError UnknownRef
+    Just _ -> throwError MissingInType -- $ WrongInType x -- TODO
```
Comment (ak3n) [inline plutus-ledger/src/Ledger/Typed/Scripts.hs:115]: Let's add comment why `TODO` and what is necessary?

File: plutus-ledger/src/Ledger/Typed/Scripts.hs [hunk 10]
```diff
@@ -1,9 +1,116 @@
...
+      && tyTxInOutRef l == tyTxInOutRef r
+
+-- instance (FromJSON (DatumType a), FromData (DatumType a), ToData (DatumType a)) => FromJSON (TypedScriptTxIn a) where
+--   parseJSON (Data.Aeson.Object v) =
+--     TypedScriptTxIn <$> v .: "tyTxInTxIn" <*> v .: "tyTxInOutRef"
+--   parseJSON invalid = typeMismatch "Object" invalid
+
+-- instance (ToJSON (DatumType a)) => ToJSON (TypedScriptTxIn a) where
+--   toJSON TypedScriptTxIn {tyTxInTxIn, tyTxInOutRef} =
+--     object ["tyTxInTxIn" .= tyTxInTxIn, "tyTxInOutRef" .= tyTxInOutRef]
```
Comment (ak3n) [inline plutus-ledger/src/Ledger/Typed/Scripts.hs:48]: Can be removed?

File: plutus-ledger/src/Ledger/Typed/Scripts.hs [hunk 11]
```diff
@@ -1,9 +1,116 @@
...
+-- | Create a 'PubKeyTxIn' from an existing 'TxIn' by checking that it has the right payment type.
+typePubKeyTxIn ::
+  forall m.
+  (MonadError ConnectionError m) =>
+  TxIn ->
+  m PubKeyTxIn
+typePubKeyTxIn txIn =
+  case txInType txIn of
+    Just ConsumePublicKeyAddress -> pure $ PubKeyTxIn txIn
+    Just _                       -> throwError MissingInType -- $ WrongInType x -- TODO
```
Comment (ak3n) [inline plutus-ledger/src/Ledger/Typed/Scripts.hs:88]: Same as below.

File: plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs [hunk 12]
```diff
@@ -631,21 +652,32 @@ processConstraint = \case
...
             unbalancedTx . tx . Tx.inputs %= (input :)
+            inputs <- use (unbalancedTx . tx . Tx.inputs)
+            -- We use fromJust because we can garanty that it will always be Just.
+            let idx = fromJust $ elemIndex input inputs
             unbalancedTx . tx . Tx.datumWitnesses . at dvh .= Just datum
+            unbalancedTx . tx . Tx.redeemers . at (RedeemerPtr Spend (fromIntegral idx)) .= Just red
             valueSpentInputs <>= provided value
           _ -> throwError (TxOutRefWrongType txo)
+    MustUseOutputAsCollateral _ -> do
+        pure () -- TODO
```
Comment (ak3n) [inline plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs:669]: Empty todo.

**Review by ak3n — APPROVED** [review]
That's a lot! Nice work! LGTM.

Mostly my comments are about empty and missing `TODO`s.

File: plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs [hunk 13]
```diff
@@ -510,6 +520,25 @@ mustSpendScriptOutputWithMatchingDatumAndValue vh datumPred valuePred red =
         txConstraintFuns = TxConstraintFuns [MustSpendScriptOutputWithMatchingDatumAndValue vh datumPred valuePred red ]
     }
 
+{-# INLINABLE mustUseOutputAsCollateral #-}
+-- | TODO
```
[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs:524]: We have a ticket to implement `mustUseOutputAsCollateral` properly, probably for next sprint. I just needed it to be able to get the test to succeed.

File: plutus-ledger/src/Ledger/Typed/Scripts.hs (hunk 10 above)
Comment (sjoerdvisscher) [reply plutus-ledger/src/Ledger/Typed/Scripts.hs]: No this still needs to be fixed, but that involves some more changes in this PR is already too big. Will make a ticket for it.

File: plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs (hunk 5 above)
Comment (sjoerdvisscher) [reply plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs:46]: "Ld" was used twice

File: plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs [hunk 14]
```diff
@@ -626,3 +655,50 @@ collectFromTheScript ::
...
+
+collectFromPlutusV2ScriptFilter
+    :: (TxOutRef -> ChainIndexTxOut -> Bool)
+    -> Map Address (Map TxOutRef ChainIndexTxOut)
+    -> Validator
+    -> Redeemer
+    -> UntypedConstraints
+collectFromPlutusV2ScriptFilter flt am vls red = -- (Redeemer red) =
+    -- let mp'  = fromMaybe mempty $ am ^. at (PV2.mkValidatorAddress vls)
+    -- in collectFromTheScriptFilter @PlutusTx.BuiltinData @PlutusTx.BuiltinData flt mp' red
```
[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs:674]: I think this is part of the TODO just below it.

File: plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs [hunk 15]
```diff
@@ -298,7 +308,6 @@ mustIncludeDatum = singleton . MustIncludeDatum
 mustPayToTheScript :: forall i o. PlutusTx.ToData o => o -> Value -> TxConstraints i o
 mustPayToTheScript dt vl =
     mempty { txOwnOutputs = [ScriptOutputConstraint dt vl] }
-    <> mustIncludeDatum (Datum (PlutusTx.toBuiltinData dt))
```
Comment (koslambrou) [inline plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs:301]: @sjoerdvisscher Why was this removed?

File: plutus-ledger/src/Ledger/Typed/Scripts.hs (hunk 10 above)
Comment (koslambrou) [reply plutus-ledger/src/Ledger/Typed/Scripts.hs]: Seems like it was removed in a recent commit. I don't see this commented code in the PR. Shouldn't we keep it?

File: plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs (hunk 15 above)
Comment (sjoerdvisscher) [reply plutus-ledger-constraints/src/Ledger/Constraints/TxConstraints.hs:301]: `ScriptOutputConstraint` already does with the datum what `mustIncludeDatum` does, both on-chain and off-chain. And for ledger-tx-constraints it doesn't make sense to split it up like this. Same for the other cases.

File: plutus-ledger/src/Ledger/Typed/Scripts.hs (hunk 10 above)
Comment (sjoerdvisscher) [reply plutus-ledger/src/Ledger/Typed/Scripts.hs]: No, after some more thought it seems we don't need it.

**Review by sjoerdvisscher — APPROVED** [review]
